File: js/src/providers/kagi.ts
import type { HttpResponse } from "../http.js";
import type { SearchRequest, SearchResponse } from "../types.js";
import { finalize, result, type ProviderContext, type ProviderSpec } from "./base.js";

export const kagi: ProviderSpec = {
  name: "kagi",
  envKeys: ["KAGI_API_KEY"],
  defaultBaseUrl: "https://kagi.com",
  capabilities: new Set([]),

  prepare(ctx: ProviderContext, req: SearchRequest) {
    const params: Record<string, string | number | boolean | undefined> = {
      q: req.query,
      limit: req.maxResults,
    };
    for (const [k, v] of Object.entries(req.extra)) params[k] = v as string;
    return {
      method: "GET",
      url: `${ctx.baseUrl}/api/v0/search`,
      headers: { Authorization: `Bot ${ctx.apiKey ?? ""}`, Accept: "application/json" },
      params,
    };
  },

  parse(_ctx, res: HttpResponse, req: SearchRequest, elapsedMs: number): SearchResponse {
    const data = res.data ?? {};
    const results = (data.data ?? [])
      .filter((item: any) => item.t === 0)
      .map((item: any) =>
        result({
          title: item.title,
          url: item.url,
          snippet: item.snippet,
          publishedDate: item.published,
          raw: item,
        }),
      );
    return finalize(kagi, req, results, data, { requestId: data.meta?.id, elapsedMs });
  },
};

File: python/src/anysearch/providers/kagi.py
"""Kagi Search API (v0) — premium, ad-free search index."""

from __future__ import annotations

import httpx

from .._http import PreparedRequest
from ..types import SearchRequest, SearchResponse
from .base import BaseProvider


class KagiProvider(BaseProvider):
    name = "kagi"
    env_keys = ("KAGI_API_KEY",)
    default_base_url = "https://kagi.com"
    capabilities = frozenset(set())

    def prepare(self, req: SearchRequest) -> PreparedRequest:
        params = {"q": req.query, "limit": req.max_results}
        params.update(req.extra)
        return PreparedRequest(
            method="GET",
            url=f"{self.base_url}/api/v0/search",
            headers={"Authorization": f"Bot {self.api_key}", "Accept": "application/json"},
            params=params,
        )

    def parse(self, response: httpx.Response, req: SearchRequest, elapsed_ms: float) -> SearchResponse:
        data = response.json()
        results = []
        for item in data.get("data", []) or []:
            # t == 0 is a search result; t == 1 is "related searches".
            if item.get("t") != 0:
                continue
            results.append(
                self._result(
                    title=item.get("title"),
                    url=item.get("url"),
                    snippet=item.get("snippet"),
                    published_date=item.get("published"),
                    raw=item,
                )
            )
        meta = data.get("meta") or {}
        return self._response(req, results, data, request_id=meta.get("id"), elapsed_ms=elapsed_ms)
